### PR TITLE
WIP: Allow deletion of pending pods when using PDBS

### LINKED ISSUE #80389: PDBs using minUnavailable prevent draining unready pods
**What happened**:

We were unable to drain pods on an unhealthy node that were stuck in pod initializing. The disruption budget has a `minAvailable` of 2. We had two healthy pods and a pod stuck in initializing thus it's unhealthy. A drain on the node with the unhealthy pod stuck in initializing was prevented. 

**What you expected to happen**:

The pods should have been evicted because they are in a not Ready state. 

**How to reproduce it (as minimally and precisely as possible)**:

Using a PDB with `minAvailable` attempt to drain a node with a pod in a not Ready state. 

**Anything else we need to know?**:

The reason this is prevented is because the disruption controller determines the number of allowed disruptions by looking at [`currentHealthy - desiredHealthy`](https://github.com/kubernetes/kubernetes/blob/a07b027261383127494fc6d4e3ca2c544fb4a464/pkg/controller/disruption/disruption.go#L749-L752). When `minAvailable` is used your `desiredHealthy` count becomes that value, which in our case is 2. The `currentHealthy` count is 2 also because the 3rd pod is stuck in an initializing state on the bad node. This leaves the PDB with an allowed disruption of 0 and thus blocks the drain. 

The eviction logic in the storage only allows pods in a terminal state to bypass the PDB. 

https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L124-L133

If they're not in a terminal state the eviction will only be allowed to occur if there are allowable disruptions available on the PDB.

It seems that the storage logic should allow draining pods that are not ready instead of just allowing terminal pods. 

/sig apps

### PR DESCRIPTION
**Pull request by michaelgugino**
Currently, if you have a PDB set, it is possible for
a pod stuck in pending state to be prevented from
deletion even though there are in fact enough healthy
replicas.

This commit allows pods in Pending state to be removed
as long as there are enough healthy pods according to
the PDB.

This commit also adds associated unit tests.

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Currently, if you have a PDB set, it is possible for
a pod stuck in pending state to be prevented from
deletion even though there are in fact enough healthy
replicas.
**Which issue(s) this PR fixes**:

Fixes: https://github.com/kubernetes/kubernetes/issues/80389

**Special notes for your reviewer**:
Here's where 'healthy' pods are counted: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/disruption/disruption.go#L675

**Does this PR introduce a user-facing change?**:

```release-note
When evicting, Pods in Pending state to be removed as long as there are enough healthy pods according PDBs.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by praseodym** [comment]
Wouldn’t it better to ignore all unready pods in the PDB healthy pods calculation? I ran into this issue for pods that were started but never passed their readiness checks.

**Comment by michaelgugino** [comment]
> Wouldn’t it better to ignore all unready pods in the PDB healthy pods calculation? I ran into this issue for pods that were started but never passed their readiness checks.

My approach is to be as conservative as possible.  This PR doesn't affect the counting of healthy pods, that is done elsewhere.  What this does is allow you to evict a particular pod if there are enough healthy pods elsewhere and the pod you want to evict is 'unready', thus not included in the 'healthy' count.

The 'healthy' count only includes 'ready' pods.  Pods can possibly be in a variety of states other than 'ready' and not be included in the healthy count; I'm not sure what all those states are, I think 'unready' is a good starting point.

**Comment by praseodym** [comment]
> > Wouldn’t it better to ignore all unready pods in the PDB healthy pods calculation? I ran into this issue for pods that were started but never passed their readiness checks.
> 
> My approach is to be as conservative as possible. This PR doesn't affect the counting of healthy pods, that is done elsewhere. What this does is allow you to evict a particular pod if there are enough healthy pods elsewhere and the pod you want to evict is 'unready', thus not included in the 'healthy' count.
> 
> The 'healthy' count only includes 'ready' pods. Pods can possibly be in a variety of states other than 'ready' and not be included in the healthy count; I'm not sure what all those states are, I think 'unready' is a good starting point.

My original comment was inaccurate, so let me rephrase: wouldn't it be better to allow any unready pod to be evicted when there are enough healthy pods?

I recently ran into an issue where I had two pods, one ready (healthy) and one unready (unhealthy, due to a node-local issue), with a pod disruption budget of minAvailable=1. Draining the node with the unready pod didn't work because the eviction of the unready pod would allegedly cause the number of available pods to become lower than minAvailable.

This means the bug is probably a little wider than just pending pods; all unhealthy pods (so not included in the [set of healthy pods](https://github.com/kubernetes/kubernetes/blob/a07b027261383127494fc6d4e3ca2c544fb4a464/pkg/controller/disruption/disruption.go#L686-L689)) should be allowed for immediate eviction.

**Comment by michaelgugino** [comment]
> > > Wouldn’t it better to ignore all unready pods in the PDB healthy pods calculation? I ran into this issue for pods that were started but never passed their readiness checks.
> > 
> > 
> > My approach is to be as conservative as possible. This PR doesn't affect the counting of healthy pods, that is done elsewhere. What this does is allow you to evict a particular pod if there are enough healthy pods elsewhere and the pod you want to evict is 'unready', thus not included in the 'healthy' count.
> > The 'healthy' count only includes 'ready' pods. Pods can possibly be in a variety of states other than 'ready' and not be included in the healthy count; I'm not sure what all those states are, I think 'unready' is a good starting point.
> 
> My original comment was inaccurate, so let me rephrase: wouldn't it be better to allow any unready pod to be evicted when there are enough healthy pods?
> 
> I recently ran into an issue where I had two pods, one ready (healthy) and one unready (unhealthy, due to a node-local issue), with a pod disruption budget of minAvailable=1. Draining the node with the unready pod didn't work because the eviction of the unready pod would allegedly cause the number of available pods to become lower than minAvailable.
> 
> This means the bug is probably a little wider than just pending pods; all unhealthy pods (so not included in the [set of healthy pods](https://github.com/kubernetes/kubernetes/blob/a07b027261383127494fc6d4e3ca2c544fb4a464/pkg/controller/disruption/disruption.go#L686-L689)) should be allowed for immediate eviction.

So, I'm not saying you're wrong.  My suggestion is to start with pending, get this reviewed, and then iterate from there.  I'm not super familiar with all the different pod states.

File: pkg/registry/core/pod/storage/eviction.go [hunk 1]
```diff
@@ -154,6 +154,22 @@ func (r *EvictionREST) Create(ctx context.Context, obj runtime.Object, createVal
 		pdb := &pdbs[0]
 		pdbName = pdb.Name
 		refresh := false
+		// Pending Pods don't count towards healthy; allow removal of pending pod as long as we have enough healthy ones.
+		if pod.Status.Phase == api.PodPending && pdb.Spec.MaxUnavailable != nil {
```
Comment (smarterclayton) [inline pkg/registry/core/pod/storage/eviction.go]: Is the bug in this case a pod that hasn’t been scheduled yet?  A pod that hasn’t been scheduled can absolutely no-questions-asked be evicted, and I don’t see that handled up above in this file.  If the bug is scheduled but not started, I’d probably still like a section above for it.

Is there any scenario where a pending pod being marked deleted could violate the constraints PDB seeks to maintain?  Ie due to a delay in propagating status?  I don’t think there is but I want to sure you’ve thought through it.

File: pkg/registry/core/pod/storage/eviction.go [hunk 2]
```diff
@@ -154,6 +154,22 @@ func (r *EvictionREST) Create(ctx context.Context, obj runtime.Object, createVal
...
+			expectedPods := int(pdb.Status.ExpectedPods)
+			maxUnavailable, err := intstr.GetValueFromIntOrPercent(pdb.Spec.MaxUnavailable, expectedPods, true)
+			if err == nil {
+				if int(pdb.Status.CurrentHealthy) >= expectedPods-maxUnavailable {
+					_, _, err = r.store.Delete(ctx, eviction.Name, rest.ValidateAllObjectFunc, deletionOptions)
```
Comment (smarterclayton) [inline pkg/registry/core/pod/storage/eviction.go]: There is no decrement necessary here?

**Comment by smarterclayton** [comment]
> This PR doesn't affect the counting of healthy pods, that is done elsewhere. What this does is allow you to evict a particular pod if there are enough healthy pods elsewhere and the pod you want to evict is 'unready', thus not included in the 'healthy' count.

Technically there is a race between health calculation and pods changing status where the healthy count could include the pod that is under eviction.  That would only be possible if pod phase could be reset, which it’s not supposed to be (but is currently possible due to another bug we are working on).  Since phase is one way, pending pods should never had previously been factored into running pod counts (can’t be ready and pending) so this should be safe.

**Comment by smarterclayton** [comment]
I’m shocked this is missing(maybe the logic caused this not to happen!, but eviction should also ignore pods that are already marked deleted and not double decrement.  It would be good to verify that constraint is being held

**Comment by smarterclayton** [comment]
It looks like double decrement is not happening, but we should probably verify that delete of an existing delete isn’t doing unnecessary work.

File: pkg/registry/core/pod/storage/eviction.go (hunk 1 above)
Comment (mortent) [reply pkg/registry/core/pod/storage/eviction.go]: I think I agree with this. If the pod is in a pending state, we don't really need to care about PDBs at all since the pod is not part of the PDB calculations. Then we should be able to handle this case above with the deletion of terminal pods.

Could there be a race where a pod reaches the running phase and is observed by the disruption controller before it gets deleted here? It does sound unlikely, but I think it might be possible.

Comment (mortent) [reply pkg/registry/core/pod/storage/eviction.go]: So I am not very concerned about the possible race here. It seems like there will always be a chance that reality changes between the check in the controller and the delete of a pod.

I'm not convinced the extra checks here are necessary. If we determine that the pod is in the pending state, I think we should be able to delete it without any additional checks since pending pods are not part of the pdb calculation. This means we should be able to just add a check for pending pods on line 126. Do you have any concerns about doing it that way?

Comment (michaelgugino) [reply pkg/registry/core/pod/storage/eviction.go]: @mortent The only thing I can really think of is that if PDBs are already below tolerance, then only checking for pending state could potentially prolong getting to the right number of replicas.  While not ideal, the application is no worse off than it was.

I'll cut a new patch for just checking on 126.

**Comment by liggitt** [comment]
/close
in favor of https://github.com/kubernetes/kubernetes/pull/83906

**Comment by michaelgugino** [comment]
/reopen

This is being discussed as potentially solving the problem of crashloopbackoff re-opening for now.

**Comment by mortent** [comment]
So the approach here to check whether a pod can be terminated is identical to how the disruption controller computes the `DisruptionsAllowed`, but it bypasses the decrements performed by the eviction api. Why is this necessary vs checking the `DisruptionsAllowed` field? And if the `DisruptionsAllowed` field is decremented incorrectly, can this be fixed there instead?

I also think this solution might have a race where if a Pod are evicted, the `CurrentHealthy` field on the PDB might not be updated before the next call to evict.

**Comment by michaelgugino** [comment]
> So the approach here to check whether a pod can be terminated is identical to how the disruption controller computes the `DisruptionsAllowed`, but it bypasses the decrements performed by the eviction api.

Not exactly.  DisruptionsAllowed is decremented by 1 per non-ready pod.  The approach is to check that the number of ready pods is greater than or equal to the minimum.

> Why is this necessary vs checking the `DisruptionsAllowed` field? And if the `DisruptionsAllowed` field is decremented incorrectly, can this be fixed there instead?

It's not decremented incorrectly.  DisruptionsAllowed works in most cases as designed.  Here's more discussion of the problem: https://github.com/kubernetes/kubernetes/issues/80389

> I also think this solution might have a race where if a Pod are evicted, the `CurrentHealthy` field on the PDB might not be updated before the next call to evict.

I think the window for this would be quite small, and would require two+ simultaneous evictions to two+ pods (1 or more going unready at the exact moment eviction is called and before PDBs can be updated).  I think we could add the resourceVersion precondition to deleting the pod as we're doing in the other patch and that would catch this issue (though it's unclear how that might play with crashloop-backoff as that's probably going to be noisy on the pod object)

**Comment by mortent** [comment]
> > So the approach here to check whether a pod can be terminated is identical to how the disruption controller computes the `DisruptionsAllowed`, but it bypasses the decrements performed by the eviction api.
> 
> Not exactly. DisruptionsAllowed is decremented by 1 per non-ready pod. The approach is to check that the number of ready pods is greater than or equal to the minimum.

DisruptionsAllowed is computed by taking the difference between the number of currently healthy pod and the desired number of healthy pods
https://github.com/kubernetes/kubernetes/blob/09cb73a554d3218f0267ba28de8a5fba47fd6335/pkg/controller/disruption/disruption.go#L764
The desired number of healthy pods is just the expected number of pods minus the maximum number of unavailable pods
https://github.com/kubernetes/kubernetes/blob/09cb73a554d3218f0267ba28de8a5fba47fd6335/pkg/controller/disruption/disruption.go#L605
This is the same computation as the right hand side of the if-statement on line 162 in this PR.

So if we are checking whether `CurrentlyHealthy` is larger than the desired number of ready pods, or if we check whether `CurrentlyHealthy` minus the desired number of ready pods is larger than 0, it is the same check as far as I can tell.

The main difference with the approach here is that the decremented value of `DisruptionsAllowed` is not used in the computation, which I think makes it possible to terminate more than one pod before the disruption controller has been able to recompute the `CurrentlyHealthy` field in the status.

> 
> > Why is this necessary vs checking the `DisruptionsAllowed` field? And if the `DisruptionsAllowed` field is decremented incorrectly, can this be fixed there instead?
> 
> It's not decremented incorrectly. DisruptionsAllowed works in most cases as designed. Here's more discussion of the problem: #80389
> 
> > I also think this solution might have a race where if a Pod are evicted, the `CurrentHealthy` field on the PDB might not be updated before the next call to evict.
> 
> I think the window for this would be quite small, and would require two+ simultaneous evictions to two+ pods (1 or more going unready at the exact moment eviction is called and before PDBs can be updated). I think we could add the resourceVersion precondition to deleting the pod as we're doing in the other patch and that would catch this issue (though it's unclear how that might play with crashloop-backoff as that's probably going to be noisy on the pod object)

I think we should move the discussion about eviction of crash-looping pods into an issue rather than in this PR. I will create one or see if there is one already. Eviction of pending pods is something we all agree on, so we should try to get your other PR merged.

**Comment by michaelgugino** [comment]
Closing in favor of: https://github.com/kubernetes/kubernetes/pull/94381
